components/script/dom/globalscope.rs

```diff
@@ -773,6 +852,127 @@ impl GlobalScope {
             .send(ScriptMsg::RerouteMessagePort(port_id, task));
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#dom-broadcastchannel-postmessage>
+    /// Step 7 and following steps.
+    pub fn schedule_broadcast(&self, msg: BroadcastMsg, channel_id: &Uuid) {
+        // First, broadcast locally.
+        self.broadcast_message_event(msg.clone(), Some(channel_id));
+
+        if let BroadcastChannelState::Managed(router_id, _) =
+            &*self.broadcast_channel_state.borrow()
+        {
+            // Second, broadcast to other globals via the constellation.
+            //
+            // Note: for globals in the same script-thread,
+            // we could skip the hop to the constellation.
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::ScheduleBroadcast(router_id.clone(), msg));
+        } else {
+            panic!("Attemps to broadcast a message via global not managing any channels.");
+        }
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-broadcastchannel-postmessage>
+    /// Step 7 and following steps.
+    pub fn broadcast_message_event(&self, event: BroadcastMsg, channel_id: Option<&Uuid>) {
+        let BroadcastMsg {
+            data,
+            origin,
+            channel_name,
+        } = event;
+
+        // Steps 7, a few preliminary steps.
+
+        // - Check the worker is not closing.
+        if let Some(worker) = self.downcast::<WorkerGlobalScope>() {
+            if worker.is_closing() {
+                return;
+            }
+        }
+
+        // - Check the associated document is fully-active.
+        if let Some(window) = self.downcast::<Window>() {
+            if !window.Document().is_fully_active() {
+                return;
+            }
+        }
+
+        let to_dispatch = if let BroadcastChannelState::Managed(_, channels) =
+            &*self.broadcast_channel_state.borrow()
+        {
+            // Step 7: Check for a case-sensitive match for the name of the channel.
+            let channel_name = DOMString::from_string(channel_name);
+            let channels: Vec<DomRoot<BroadcastChannel>> = match channels.get(&channel_name) {
+                Some(channels) => channels
+                    .iter()
+                    .filter(|ref channel| {
+                        // Step 8.
+                        // Filter out the sender.
+                        if let Some(id) = channel_id {
+                            !(channel.id() == id)
+                        } else {
+                            true
+                        }
+                    })
+                    .map(|channel| DomRoot::from_ref(&**channel))
+                    // Step 9, sort by creation order, done by using a queue.
```

Comment:
How is this sorting? We're collecting into a vec

Reply:
Oh yeah, I'll change the target of the `collect` to be a `VecDeque`...

Reply:
Is that going to sort it in any way? That's just going to give it to you in iteration order, and we immediately iterate so the collect may not be necessary? Is it not already in creation order?

Reply:
Ah yes, so the original data structure is a queue, in which we store the channels as we create them, so the iterator will be in creation order, oldest first. Will remove the `collect`.